collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryViewModel.java

```diff
@@ -139,31 +154,75 @@ public boolean canAddRepeat() {
         }
     }
 
-    public void moveForward() {
-        try {
-            formController.stepToNextScreenEvent();
-        } catch (JavaRosaException e) {
-            error.setValue(new NonFatal(e.getCause().getMessage()));
-            return;
-        }
+    public void moveForward(HashMap<FormIndex, IAnswerData> answers) {
+        moveForward(answers, false);
+    }
+
+    public void moveForward(HashMap<FormIndex, IAnswerData> answers, Boolean evaluateConstraints) {
+        isLoading.setValue(true);
+
+        scheduler.immediate((Supplier<Boolean>) () -> {
```

Comment:
You can clean this you don't need `Supplier<Boolean>` and `return`. The same below in `moveBackward`.

Reply:
How about I just convert to Kotlin once you're happy with everything in the PR? Wanted to make sure it was easier to review the individual changes but no reason not to convert once you've been over everything.

Reply:
Sounds good.

Reply:
Sorry should have said: will do that at the end just to keep the diff easier to read (for myself as well).